torchmultimodal/modules/encoders/swin_transformer_3d_encoder.py

```diff
@@ -0,0 +1,518 @@
+    relative_position_bias: Tensor,
+    window_size: Tuple[int, int, int],
+    num_heads: int,
+    shift_size: Tuple[int, int, int] = (0, 0, 0),
+    attention_dropout: float = 0.0,
+    dropout: float = 0.0,
+    qkv_bias: Optional[Tensor] = None,
+    proj_bias: Optional[Tensor] = None,
+):
+    """
+    Window based multi-head self attention (W-MSA) module with relative position bias.
+    It supports both of shifted and non-shifted window.
+    Args:
+        input (Tensor[B, D, H, W, C]): The input tensor, 5-dimensions.
+        qkv_weight (Tensor[in_dim, out_dim]): The weight tensor of query, key, value.
+        proj_weight (Tensor[out_dim, out_dim]): The weight tensor of projection.
+        relative_position_bias (Tensor): The learned relative position bias added to attention.
+        window_size (Tuple[int, int, int]): 3-dimensions window size, D, H, W .
+        num_heads (int): Number of attention heads.
+        shift_size (Tuple[int, int, int]): Shift size for shifted window attention (D, H, W). Default: (0, 0, 0)
+        attention_dropout (float): Dropout ratio of attention weight. Default: 0.0.
+        dropout (float): Dropout ratio of output. Default: 0.0.
+        qkv_bias (Tensor[out_dim], optional): The bias tensor of query, key, value. Default: None.
+        proj_bias (Tensor[out_dim], optional): The bias tensor of projection. Default: None.
+    Returns:
+        Tensor[B, D, H, W, C]: The output tensor after shifted window attention.
+    """
+    B, D, H, W, C = input.shape
+    # pad feature maps to multiples of window size
+    pad_size = _compute_pad_size_3d((D, H, W), window_size)
+    x = F.pad(input, (0, 0, 0, pad_size[2], 0, pad_size[1], 0, pad_size[0]))
+    _, Dp, Hp, Wp, _ = x.shape
+    padded_size = [Dp, Hp, Wp]
+
+    # cyclic shift
+    if sum(shift_size) > 0:
+        x = torch.roll(
+            x, shifts=(-shift_size[0], -shift_size[1], -shift_size[2]), dims=(1, 2, 3)
+        )
+
+    # partition windows
+    num_windows = (
+        (padded_size[0] // window_size[0])
+        * (padded_size[1] // window_size[1])
+        * (padded_size[2] // window_size[2])
+    )
+    x = x.view(
+        B,
+        padded_size[0] // window_size[0],
+        window_size[0],
+        padded_size[1] // window_size[1],
+        window_size[1],
+        padded_size[2] // window_size[2],
+        window_size[2],
+        C,
+    )
+    x = x.permute(0, 1, 3, 5, 2, 4, 6, 7).reshape(
+        B * num_windows, window_size[0] * window_size[1] * window_size[2], C
+    )  # B*nW, Wd*Wh*Ww, C
+
+    # multi-head attention
+    qkv = F.linear(x, qkv_weight, qkv_bias)
+    qkv = qkv.reshape(x.size(0), x.size(1), 3, num_heads, C // num_heads).permute(
+        2, 0, 3, 1, 4
+    )
+    q, k, v = qkv[0], qkv[1], qkv[2]
+    q = q * (C // num_heads) ** -0.5
+    attn = q.matmul(k.transpose(-2, -1))
+    # add relative position bias
+    attn = attn + relative_position_bias
+
+    if sum(shift_size) > 0:
```

Comment:
Can we add a test when this is not true?

Reply:
sure

Reply:
added the test on `test/modules/encoder/test_swin_transformer_3d_encoder.py:TestSwinTransformer3dComponents.test_shifted_window_attention_3d_zero_shift`